### ISSUE TITLE
Task 2.4: Research Flask

### ISSUE THREAD

**Issue by joshuag1214** [joshuag1214/ENPM613-Project#18]

**Comment by joshuag1214** [comment]
https://flask.palletsprojects.com/en/2.3.x/tutorial/

**Comment by thefueley** [comment]
Tacking my [proof of concept repo](https://github.com/thefueley/sonic-poc) here in case anyone wants to see. I still need to finish the tutorial linked in the task, but what is finished right now is: Docker build / push, makefile to run in local Docker, Azure DevOps CI/CD pipeline. What remains: remaining bit of tutorial, Docker ignores, configuring the remaining stages in the CI/CD pipeline, and demo.

**Comment by joshuag1214** [comment]
For security, I imagine that the TA/professor cares about if the user is able to access another user's data/permission, EX: student can't create classes. We can ask TA/Prof for their opinions on TLS certificates?

**Comment by joshuag1214** [comment]
Some vids related to Chris's resource that I found helpful: 
https://www.youtube.com/watch?v=CZYY0fw1Wp0
https://www.youtube.com/watch?v=AgVqsmz-ZW4&list=

**Comment by thefueley** [comment]
@joshuag1214, I agree that the professor cares about the basic security case you laid out, but I also think the security requirements are open-ended. I feel that there are just a _few hard_ requirements for the project. The prof/ta _mostly_ only care about those. We're free to add or remove other features. That's how I've interpreted the Q&A back and forth with the prof/ta.

The _Software Dev Project Description_ says, "_The main quality attributes that will drive your software will be usability and security. You will need to protect system data and services and build in security from the beginning, by identifying security requirements, as well as by applying adequate architecture, detailed design, and implementation practices._"

Feel free to reach out and ask them if we _have to_ use TLS, but I can't picture any scenario in computing where authentication is required and encryption is not, even in a test environment. Even if the prof/ta responds saying it isn't required, I think we should still do it. Our very first use case, **UC001 Login User**, is where TLS applies. While TLS isn't a fix-all for security, without it our LMS is open to many types of attacks, like Man-in-Middle, session hijacking, and altering of data since everything is just plain text. I don't see how we could say we built this app with security in mind but blatantly ignore encryption. My opinion is that TLS is mandatory. 

If complexity/labor to set up TLS is your concern, I am willing to show the team how to accomplish this. My thought it to use Let's Encrypt, which automates the maintenance of TLS and is free.

**Comment by joshuag1214** [comment]
Ok that's fine then. I thought you would need to pay for something to get
the TLS certificate

**Comment by joshuag1214** [comment]
commands to run repo locally 

flask --app sonic run --debug

flask --app sonic  init-db

**Comment by thefueley** [comment]
The PoC is configured to run as a container. There is a makefile in the repo. So if you install Docker locally, you develop locally and run the app just as it will run in production. Install Docker, code away, the run 'make build'. This builds the container. Run the code with 'make run'. Code some more and make the changes you want. Run 'make build' to build those new code changes into a new container. Then run 'make rerun' to run the new code. Visit http://localhost:5001 to see the app in a browser

**Comment by joshuag1214** [comment]
Not to familiar with sql/.sqlite but used this: https://sqlitebrowser.org/ 
to view sonic-poc-master/instance/flaskr.sqlite

**Comment by joshuag1214** [comment]
another useful resource:https://www.youtube.com/watch?v=v3CSQkPJtAc
